### PR TITLE
Update and add dev guide content

### PR DESCRIPTION
**Pull request by rhornung67**
# Summary

- This PR adds content to the RAJA Dev Guide about CI test organization and maintenance tasks.

The generated documentation can be viewed here: https://raja.readthedocs.io/en/task-ci-docs/sphinx/dev_guide/index.html

### CODE REVIEW

**Comment by kab163** [comment]
Is there a link to the readthedocs that this generates?

**Review by kab163 — COMMENTED** [review]
Should there be any mentioning of how to create a new spec job? (Or updating a job if updates are done to a LC machine to where a spec no longer works?) That may require adding certain flags in the `scripts/spack_packages/raja/package.py` file or adding other parameters to the spec like `cxxflags=...`

**Comment by rhornung67** [comment]
> Should there be any mentioning of how to create a new spec job? (Or updating a job if updates are done to a LC machine to where a spec no longer works?) That may require adding certain flags in the `scripts/spack_packages/raja/package.py` file or adding other parameters to the spec like `cxxflags=...`

Yes. I am planning to add a section on CI maintenance tasks that will cover those sorts of things.

File: docs/sphinx/dev_guide/build_configurations.rst [hunk 1]
```diff
@@ -113,18 +117,18 @@ in sub-directories in the RAJA ``scripts/uberenv/spack_configs`` directory:
   config.yaml
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/build_configurations.rst:117]: that section is old and needs to be updated...

File: docs/sphinx/dev_guide/build_configurations.rst [hunk 2]
```diff
@@ -135,8 +139,7 @@ the compiler and version. For example,
   hc-quartz-toss_3_x86_64_ib-gcc@8.1.0-fjcjwd6ec3uen5rh6msdqujydsj74ubf.cmake
 
 Specs that are exercised during the Gitlab CI process are found YAML files in 
-the ``RAJA/.gitlab`` directory. See :ref:`vettedspecs-label` for more 
-information.
+the ``RAJA/.gitlab`` directory. See :ref:`ci-label` for more information.
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/build_configurations.rst:142]: ditto.

File: docs/sphinx/dev_guide/ci.rst [hunk 3]
```diff
@@ -12,78 +12,329 @@
...
+--------------------------------------
+
+The following figure shows directories and files in the RAJA repo that 
+support LC CZ Gitlab CI testing. Files with names in blue are specific to RAJA 
+and are maintained by the RAJA team. Directories and files with names in red are
+in Git submodules, shared and maintained with other projects.
+
+.. figure:: ./figures/RAJA-Gitlab-Files.png
+
+   Directories and files in the RAJA repo that support Gitlab CI testing.
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/ci.rst:98]: That's the figure caption.

File: docs/sphinx/dev_guide/ci.rst [hunk 4]
```diff
@@ -12,78 +12,329 @@
...
+    - local: .gitlab/ruby-jobs.yml
+    - local: .gitlab/lassen-templates.yml
+    - local: .gitlab/lassen-jobs.yml
+    - local: .gitlab/corona-templates.yml
+    - local: .gitlab/corona-jobs.yml
+
+In the ``stages`` section above, prefixes 'r_', 'l_', and 'c_' refer to 
+resources in the LC on which tests are run. Specifically, the machines 'ruby',
+'lassen' and 'corona'. The jobs that will run in the pipeline(s) for each 
+resource are defined in the files listed in the ``include`` section.
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/ci.rst:136]: Dunno. I must have been sleepy.... I will fix.

File: docs/sphinx/dev_guide/ci.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (kab163) [reply docs/sphinx/dev_guide/ci.rst:98]: Perhaps "This figure shows the directories and files..."

**Comment by kab163** [comment]
I will think about this over the weekend. If I think about other things that could be added here, I'll let you know. So far, this looks like very helpful, much needed documentation. Thanks Rich!

File: docs/sphinx/dev_guide/build_configurations.rst [hunk 5]
```diff
@@ -97,11 +100,12 @@ Generating a RAJA host-config file
 
 This section describes the host-config file generation process for RAJA.
 
-Machine specific configurations
+Platform configurations
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Compiler configurations for Livermore computer platforms are contained in 
-in sub-directories in the RAJA ``scripts/uberenv/spack_configs`` directory:
+in sub-directories in the ``RAJA/scripts/uberenv/spack_configs`` directory.
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/build_configurations.rst:107]: fixed.

File: docs/sphinx/dev_guide/build_configurations.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/build_configurations.rst:117]: fixed.

File: docs/sphinx/dev_guide/build_configurations.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/build_configurations.rst:142]: fixed.

File: docs/sphinx/dev_guide/build_configurations.rst [hunk 6]
```diff
@@ -164,29 +167,34 @@ MacOS
 ^^^^^
 
 In RAJA, the Spack configuration for MacOS contains the default compiler
-corresponding to the OS version (`compilers.yaml`), and a commented section to 
-illustrate how to add `CMake` as an external package. You may install CMake 
+corresponding to the OS version in the ``compilers.yaml`` file in the 
+``RAJA/scripts/radiuss-spack-configs/darwin/`` directory, and a commented 
+section to illustrate how to add `CMake` as an external package in the
+``packages.taml`` in the same directory. You may also install CMake
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/build_configurations.rst:173]: fixed.

File: docs/sphinx/dev_guide/ci.rst [hunk 7]
```diff
@@ -12,78 +12,329 @@
...
+in the RAJA repo, not for PRs for a branch on a fork of the repo. An alternative
+procedure we use to handle this is described in :ref:`contributing-label`.
+
+Gitlab CI (LC CZ) Testing Workflow
+--------------------------------------
+
+The figure below shows the high-level steps in the RAJA Gitlab CI testing 
+process. The main steps, which we will discuss in more detail later, are:
+
+  #. *Mirror* the RAJA GitHub repo to the RAJA project in the Gitlab instance.
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/ci.rst:66]: I address your first concern. 

I'm not clear on what to say about mirroring. If you have any thoughts to share about what someone unfamiliar with it should know, please let me know, or feel free to edit the docs.

File: docs/sphinx/dev_guide/ci.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/ci.rst:98]: fixed.

File: docs/sphinx/dev_guide/ci.rst [hunk 8]
```diff
@@ -12,78 +12,329 @@
...
+    - local: .gitlab/ruby-templates.yml
+    - local: .gitlab/ruby-jobs.yml
+    - local: .gitlab/lassen-templates.yml
+    - local: .gitlab/lassen-jobs.yml
+    - local: .gitlab/corona-templates.yml
+    - local: .gitlab/corona-jobs.yml
+
+In the ``stages`` section above, prefixes 'r_', 'l_', and 'c_' refer to 
+resources in the LC on which tests are run. Specifically, the machines 'ruby',
+'lassen' and 'corona'. The jobs that will run in the pipeline(s) for each
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/ci.rst:135]: done.

File: docs/sphinx/dev_guide/ci.rst [hunk 9]
```diff
@@ -12,78 +12,329 @@
...
+    - local: .gitlab/lassen-jobs.yml
+    - local: .gitlab/corona-templates.yml
+    - local: .gitlab/corona-jobs.yml
+
+In the ``stages`` section above, prefixes 'r_', 'l_', and 'c_' refer to 
+resources in the LC on which tests are run. Specifically, the machines 'ruby',
+'lassen' and 'corona'. The jobs that will run in the pipeline(s) for each 
+resource are defined in the files listed in the ``include`` section. 
+
+The ``RAJA/.gitlab`` directory contains a *resource* and *jobs* file for each
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/dev_guide/ci.rst:138]: fixed.

**Comment by rhornung67** [comment]
@kab163 I addressed all your comments, except for adding more info about Gitlab mirroring. I also, added links to files in the repo that may help folks navigate more easily.

**Comment by rhornung67** [comment]
@mdavis36 if you have some time, please take a crack at filling in some content in the Azure CI maintenance tasks section in the dev guide, specifically, w.r.t. to the docker images we use for the builds.

**Comment by MrBurmark** [comment]
This looks good. I learning a lot about the CI config and spack stuff. I also liked the section about writing a test.

**Comment by rhornung67** [comment]
Thanks @MrBurmark. I will take another clean up pass and add a few missing things that I recently learned. Then, I will remove draft status and it should be good to go.

**Review by artv3 — DISMISSED** [review]

**Review by kab163 — DISMISSED** [review]

**Comment by rhornung67** [comment]
Thanks for the approvals. Anyone else want to review and approve before I merge?

**Comment by rchen20** [comment]
Taking a quick look now.

File: docs/sphinx/dev_guide/ci_tasks.rst [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+    **CUDA** for the cuda runtime.::
+
+      . /opt/spack/share/spack/setup-env.sh && spack load cuda
+
+    **HIP** for the hip runtime and llvm-amdgpu runtime libraries.::
+
+      . /opt/spack/share/spack/setup-env.sh && spack load hip llvm-amdgpu
+
+    **Sycl** requiers us to run setupvars.sh::
```
Comment (rchen20) [inline docs/sphinx/dev_guide/ci_tasks.rst:190]:
```suggestion
    **Sycl** requires us to run setupvars.sh::
```

File: docs/sphinx/dev_guide/release_process.rst [hunk 11]
```diff
@@ -24,19 +31,19 @@ The RAJA release process typically involves the following sequence of steps:
      into the **main branch.** When it is approved and all CI checks pass,
```
Comment (rchen20) [inline docs/sphinx/dev_guide/release_process.rst:31]: Just to clarify, we create a release candidate branch from `develop`, but merge it into `main`?

File: docs/sphinx/dev_guide/release_process.rst [hunk 12]
```diff
@@ -134,3 +146,36 @@ the key steps for performing a hotfix release are:
...
+
+After a RAJA release is complete, other tasks are performed to update content 
+in other repositories, typically. These tasks include:
+
+  * Update the `RAJAProxies <https://github.com/LLNL/RAJAProxies>`_ project
+    to the newly RAJA Portability Suite projects. This typically consists of 
+    updating the submodules to the the new RAJA Portability Suite project 
+    versions, making sure the proxy-apps build and run correctly. When this
+    is done, tag a release for proxy-app project.
+  * Update the `RAJAProxies <https://github.com/LLNL/RAJAProxies>`_ project
```
Comment (rchen20) [inline docs/sphinx/dev_guide/release_process.rst:164]: RAJAProxies mentioned twice, although both statements are slightly different.

File: docs/sphinx/dev_guide/ci_tasks.rst (hunk 10 above)
Comment (rhornung67) [reply docs/sphinx/dev_guide/ci_tasks.rst:190]: fixed.

File: docs/sphinx/dev_guide/release_process.rst (hunk 11 above)
Comment (rchen20) [reply docs/sphinx/dev_guide/release_process.rst:31]: Thanks, so `main` only has updates up until the latest release.

File: docs/sphinx/dev_guide/release_process.rst [hunk 13]
```diff
@@ -134,3 +146,32 @@ the key steps for performing a hotfix release are:
...
+=========================
+Post-release Activities
+=========================
+
+After a RAJA release is complete, other tasks are performed to update content 
+in other repositories, typically. These tasks include:
+
+  * Update the `RAJAProxies <https://github.com/LLNL/RAJAProxies>`_ project
+    to the newly RAJA Portability Suite projects. This typically consists of 
+    updating the submodules to the the new RAJA Portability Suite project
```
Comment (rchen20) [inline docs/sphinx/dev_guide/release_process.rst:161]:
```suggestion
    updating the submodules to the new RAJA Portability Suite project 
```

**Review by rchen20 — DISMISSED** [review]

File: docs/sphinx/dev_guide/release_process.rst (hunk 11 above)
Comment (rhornung67) [reply docs/sphinx/dev_guide/release_process.rst:31]: that is correct.

**Comment by rhornung67** [comment]
Folks, I took a last pass and fixed several typos, etc. unfortunately, this dismissed your approvals. Please re-approve and I'll merge this when it comes back green.

**Review by rchen20 — APPROVED** [review]

**Pull request merged**